### ISSUE TITLE
Integration tests flaking, timing out waiting for broker to come up

### ISSUE THREAD

**Issue by kibbles-n-bytes** [kubernetes-incubator/service-catalog#665]
Opening a separate issue for this flake, as #643 started to become a general hub for all flaky behavior.

Integration tests are intermittently failing with error:

```
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (40.09s)
[Build & Unit/Integration Tests]     controller_test.go:113: error waiting for broker to become ready: timed out waiting for the condition
```

**Comment by kibbles-n-bytes** [comment]
Can be seen in Jenkins builds [#173](https://service-catalog-jenkins.appspot.com/job/service-catalog-PR-testing2/173/console) and [#178](https://service-catalog-jenkins.appspot.com/job/service-catalog-PR-testing2/178/console).

**Comment by MHBauer** [comment]
```
[Build & Unit/Integration Tests] I0407 17:04:32.483384    3217 request.go:991] Response Body: {"kind":"Broker","apiVersion":"servicecatalog.k8s.io/v1alpha1","metadata":{"name":"test-broker","selfLink":"/apis/servicecatalog.k8s.io/v1alpha1/brokerstest-broker","uid":"33296edb-1bb4-11e7-8034-0242ac110002","resourceVersion":"18","creationTimestamp":"2017-04-07T17:04:02Z","finalizers":["kubernetes"]},"spec":{"url":"https://example.com"},"status":{"conditions":[{"type":"Ready","status":"False","reason":"ErrorSyncingCatalog","message":"Error syncing catalog from Broker"}]}}
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (40.10s)
[Build & Unit/Integration Tests] 	controller_test.go:113: error waiting for broker to become ready: timed out waiting for the condition
[Build & Unit/Integration Tests] 	framework.go:58: Shutting down server on port: 37268
```
It'll be nice to have more errors in the conditions to have a bit more info.

Going by the `ErrorSyncingCatalog` we've got the catalog, but can't parse it, or can't reconcile it.

**Comment by MHBauer** [comment]
Which is very confusing, because at no point are we actually running a broker, right? It's all a mock broker.

**Comment by pmorie** [comment]
Correct, and it is confusing me too. Whittled on this last night for quite a while to no effect.

**Comment by pmorie** [comment]
Observed in https://travis-ci.org/kubernetes-incubator/service-catalog/builds/219893348

**Comment by arschles** [comment]
cc/ @rifung

**Comment by arschles** [comment]
A few notes here:

- As @MHBauer said, we are not actually running a broker server
- The error shown is strictly about the success condition not showing up on the `Broker` resource

I believe this error is now fixed. Here is my hypothesis on what was happening:

1. the secure catalog API server started up, but the insecure port is 8080
2. because the insecure server couldn't serve on 8080, the `k8s.io/apiserver` package tried to restart both the secure & insecure servers (I have verified that this behavior happens in practice)
3. before it restarted, the `Broker` resource was submitted and the controller-manager picked it up and began reconciling it
4. the catalog API server shut down
5. the controller-manager hit the fake CF broker's catalog endpoint, then called the catalog API server to update the ready condition
6. because the catalog API server was not running at the time of the condition update, the ready condition was never set on the `Broker` resource
7. the controller test waited for 30 seconds (`wait.ForeverTestTimeout`) for the ready condition to be set, but it never was

Note that (1) has been fixed by @krancour's #695. This ensures the the API server can always run successfully.

I have pulled that change down, run integration tests 5 times on my machine, and haven't been able to repro. Before that change, I was able to repro this issue about once every 5 runs, so #695 has likely at least reduced the frequency of this flake.

**Comment by pmorie** [comment]
I don't think this is fixed -- I observed a race in travis with a PR rebased onto #695, but sadly have lost the link.

**Comment by pmorie** [comment]
I don't think this is fixed -- I observed a race in travis with a PR rebased onto #695, but sadly have lost the link.

**Comment by arschles** [comment]
Ok, any additional information (the branch, PR, or travis link if you can find it) would be helpful here. Leaving this open.

**Comment by pmorie** [comment]
Okay, here's another instance of the flake from #682: https://travis-ci.org/kubernetes-incubator/service-catalog/builds/221799879

**Comment by pmorie** [comment]
I have some ideas about how to fix this one, gonna hack on it after I land some other stuff

**Comment by rifung** [comment]
I think the issue is this:

> [Build & Unit/Integration Tests] I0407 17:04:02.620753    3217 request.go:991] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"serviceclasses \"test-service\" already exists","reason":"AlreadyExists","details":{"name":"test-service","kind":"serviceclasses"},"code":409}

It seems that maybe we reuse the service class "test-service" in multiple tests. I don't know enough about the test code but I imagine we're either running them in parallel or the delete is not being processed before we try creating a broker again.

Or I could be totally wrong but perhaps someone who knows better can chime in?

**Comment by arschles** [comment]
@rifung I believe that @pmorie had a similar hypothesis. Note that this behavior may be happening because the controller repeatedly fetches the catalog from the broker. After each fetch, it then tries to write service classes. If that's the case, this is a race.
